kong/clustering/services/sync/rpc.lua

```diff
@@ -0,0 +1,324 @@
+
+        local ns_delta
+
+        for namespace, delta in pairs(ns_deltas) do
+          if namespace == "default" then
+            ns_delta = delta
+            break   -- should we break here?
+          end
+        end
+
+        if not ns_delta then
+          return nil, "default namespace does not exist inside params"
+        end
+
+        if #ns_delta.deltas == 0 then
+          ngx_log(ngx_DEBUG, "no delta to sync")
+          return true
+        end
+
+        local t = txn.begin(512)
+
+        if ns_delta.wipe then
+          t:db_drop(false)
+        end
+
+        local db = kong.db
+
+        local version = 0
+        local crud_events = {}
+        local crud_events_n = 0
+
+        for _, delta in ipairs(ns_delta.deltas) do
+          local delta_type = delta.type
+          local delta_row = delta.row
+          local ev
+
+          if delta_row ~= ngx_null then
+            -- upsert the entity
+            -- does the entity already exists?
+            local old_entity, err = db[delta_type]:select(delta_row)
+            if err then
+              return nil, err
+            end
+
+            local crud_event_type = "create"
+
+            if old_entity then
+              local res, err = delete_entity_for_txn(t, delta_type, old_entity, nil)
+              if not res then
+                return nil, err
+              end
+
+              crud_event_type = "update"
+            end
+
+            local res, err = insert_entity_for_txn(t, delta_type, delta_row, nil)
+            if not res then
+              return nil, err
+            end
+
+            ev = { delta_type, crud_event_type, delta_row, old_entity, }
+
+          else
+            -- delete the entity
+            local old_entity, err = kong.db[delta_type]:select({ id = delta.id, }) -- TODO: composite key
+            if err then
+              return nil, err
+            end
+
+            if old_entity then
+              local res, err = delete_entity_for_txn(t, delta_type, old_entity, nil)
+              if not res then
+                return nil, err
+              end
+            end
+
+            ev = { delta_type, "delete", old_entity, }
+          end
+
+          crud_events_n = crud_events_n + 1
+          crud_events[crud_events_n] = ev
+
+          -- XXX TODO: could delta.version be nil or ngx.null
+          if type(delta.version) == "number" and delta.version ~= version then
+            version = delta.version
+          end
+        end
+
+        t:set(DECLARATIVE_HASH_KEY, fmt("%032d", version))
+        local ok, err = t:commit()
+        if not ok then
+          return nil, err
+        end
+
+        if ns_delta.wipe then
+          kong.core_cache:purge()
+          kong.cache:purge()
+
+        else
+          for _, event in ipairs(crud_events) do
+            -- delta_type, crud_event_type, delta.row, old_entity
+            db[event[1]]:post_crud_event(event[2], event[3], event[4])
+          end
+        end
```

Comment:
It's not clear to me why purging the cache and executing the event handlers are mutually-exclusive outcomes. Can you explain why we skip the crud events when `ns_delta.wipe` is truthy?

Reply:
We do not want to wipe the cache with incremental unless it is absolutely necessary, wiping the cache is an expensive operation and has created lots of latency issue for us previously.

Reply:
I understand why we don't purge the cache unless explicitly required, but why does this code _also_ skip crud events when a cache wipe is needed? I would assume that executing the `post_crud_event` handlers should be necessary in all cases, including when we also need to purge cache. Is that wrong?

Reply:
@flrgh I do not believe we have a way to properly execute the `crud` events with a full cache wipe, and even if we do, it might be prohibitively expensive.

Usually, `crud` event's action is to purge a specific item from the cache based on the input. It is entirely possible they might cause more side effect than that (e.g. create a timer), but that is completely out of our control.

This is also the reason why DB-less uses the `reconfigure` event instead of the `crud` event, the established pattern seems to be to handle both in plugin code. What we are doing here is not perfect, but I can't think of anything better yet.

This is listed as the second issue under: https://docs.google.com/document/d/1USZV1Hnka1o0mXNQrKTzcTIVCtLxkJwsyuo-qcqbLDs/edit?usp=sharing